### PR TITLE
Demographic model recombination rate

### PR DESCRIPTION
**Pull request by gregorgorjanc**
Here is implementation for https://github.com/popsim-consortium/stdpopsim/issues/1225.

I get almost all tests passing on my end:
* had to change `np.NINF` to `-np.inf` (new numpy)
* `TestRegisterQCDFE.test_register_qc` seems to have deep recursion, but that is not due to my changes (I think)

```
============================================================================== FAILURES ==============================================================================
_________________________________________________________________ TestRegisterQCDFE.test_register_qc _________________________________________________________________
[gw2] darwin -- Python 3.12.7 /opt/homebrew/Caskroom/miniconda/base/envs/popsim/bin/python

self = <tests.test_dfes.TestRegisterQCDFE object at 0x1094932f0>

    def test_register_qc(self):
        dfe = self.make_dfe("test")
        dfe.register_qc(dfe)
>       assert dfe.qc_dfe == dfe

tests/test_dfes.py:720:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
<attrs generated eq stdpopsim.dfe.DFE>:11: in __eq__
    self.qc_dfe == other.qc_dfe
<attrs generated eq stdpopsim.dfe.DFE>:11: in __eq__
    self.qc_dfe == other.qc_dfe
E   RecursionError: maximum recursion depth exceeded
!!! Recursion detected (same locals & position)
====================================================================== short test summary info =======================================================================
FAILED tests/test_dfes.py::TestRegisterQCDFE::test_register_qc - RecursionError: maximum recursion depth exceeded
==================================================================== 1 failed, 56 passed in 1.02s ====================================================================
```

### CODE REVIEW

File: docs/development.rst [hunk 1]
```diff
@@ -1167,6 +1165,16 @@ The first seven attributes are quite straightforward:
   However, note that this is quite uncommon, so you should make sure this is the case
   before you set the mutation rate to ``None``.
 
+* ``recombination_rate`` (`double`): The recombination rate assumed during the inference
+  of this demographic model (per bp per generation).
```
Comment (petrelharp) [inline docs/development.rst:1169]:
```suggestion
  of this demographic model, if any (per bp per generation).
```

File: docs/tutorial.rst [hunk 2]
```diff
@@ -522,18 +530,20 @@ based on the average recombination rate for that chromosome.
    # model mutation rate: 2.35e-08
 
 The Gutenkunst OOA model was inferred using a mutation rate much larger than the
-default mutation rate in the `stdpopsim` catalog. As such, simulating using this
-model and default rate will result in levels of diversity substantially lower than
-expected for the human population data that this model was inferred from. To match
-observed diversity in humans, we should instead use the mutation rate associated
-with the demographic model:
+default mutation rate for human species in the ``stdpopsim`` catalog. As such,
```
Comment (petrelharp) [inline docs/tutorial.rst:533]:
```suggestion
default mutation rate for humans in the ``stdpopsim`` catalog. As such,
```

File: stdpopsim/cli.py [hunk 3]
```diff
@@ -750,11 +759,13 @@ def run_simulation(args):
         contig = species.get_contig(
             args.chromosome,
             genetic_map=args.genetic_map,
+            # TODO: what if we would have model.genetic_map?
```
Comment (petrelharp) [inline stdpopsim/cli.py:762]: I don't think we need this comment here, and will cross this bridge if it comes up.

Comment (petrelharp) [reply stdpopsim/cli.py:762]:
```suggestion
```

File: stdpopsim/engines.py [hunk 4]
```diff
@@ -118,11 +118,28 @@ def _warn_mutation_rate_mismatch(self, contig, demographic_model):
             demographic_model.mutation_rate, contig.mutation_rate
         ):
             warnings.warn(
-                "The demographic model has mutation rate "
+                "The demographic model has mean mutation rate "
```
Comment (petrelharp) [inline stdpopsim/engines.py:121]: Demographic models (and, contigs) can't have mutation rate maps, so this is a bit misleading?
```suggestion
                "The demographic model has mutation rate "
```

File: stdpopsim/engines.py [hunk 5]
```diff
@@ -118,11 +118,28 @@ def _warn_mutation_rate_mismatch(self, contig, demographic_model):
...
+        #       demographic_model.recombination_rate?
+        if demographic_model.recombination_rate is not None and not math.isclose(
+            demographic_model.recombination_rate, contig.recombination_map.mean_rate
+        ):
+            warnings.warn(
+                "The demographic model has mean recombination rate "
+                f"{demographic_model.recombination_rate}, but this simulation used the "
+                "contig's mean recombination rate "
+                f"{contig.recombination_map.mean_rate}. "
+                "Diversity levels may be different than expected for this species. "
```
Comment (petrelharp) [inline stdpopsim/engines.py:141]:
```suggestion
                "Patterns of linkage may be different than expected for this species. "
```

File: stdpopsim/genetic_maps.py [hunk 6]
```diff
@@ -13,7 +14,7 @@
 class GeneticMap:
     """
     Class representing a genetic map for a species. Provides functionality for
-    downloading and cacheing recombination maps from a remote URL.
+    downloading and cacheing recombination/genetic maps from a remote URL.
```
Comment (petrelharp) [inline stdpopsim/genetic_maps.py:17]: According to the docs above, "genetic map" means a crossing-over map, which is what this is, so no need for the "recombination/genetic" map?
```suggestion
    downloading and cacheing genetic maps from a remote URL.
```

File: tests/test_models.py [hunk 7]
```diff
@@ -148,6 +154,8 @@ def test_all_instances(self, model):
         assert model.generation_time > 0
         if model.mutation_rate is not None:
             assert model.mutation_rate > 0
+        if model.recombination_rate is not None:
+            assert model.recombination_rate > 0
```
Comment (petrelharp) [inline tests/test_models.py:158]:
```suggestion
            assert model.recombination_rate >= 0
```

**Comment by petrelharp** [comment]
Dang, this is a big one, @gregorgorjanc! I had a careful read through and only saw a few very minor fixes.

Also of note here - clarifying terminology around "genetic map" and "recombination map"; where now it's said that "genetic map" = crossovers, while "recombination map" = crossovers and optionally GC events also.

**Comment by petrelharp** [comment]
Looks like the only thing that isn't covered by tests is in the CLI? So, just needs something to `tests/test_cli.py`? Hm, I see you *have* got something in `tests/test_cli.py`, and I don't see why that isn't doing it.

**Comment by petrelharp** [comment]
So, this *looks* good to go, once that CLI test thing is figured out - are you able to do that, @gregorgorjanc?

File: stdpopsim/engines.py (hunk 4 above)
Comment (gregorgorjanc) [reply stdpopsim/engines.py:121]: Noted. I was following how the code is setup and agree this was confusing/misleading me as well ...

File: tests/test_models.py (hunk 7 above)
Comment (gregorgorjanc) [reply tests/test_models.py:158]: Good catch!

**Comment by gregorgorjanc** [comment]
@petrelharp thanks for the review. I accepted all your suggested edits.

Will look into tests.

File: tests/test_cli.py [hunk 8]
```diff
@@ -1016,11 +1037,14 @@ def verify_noQC_citations_not_written(self, cmd, caplog):
...
         log_output = caplog.text
+        # citations are written out via the logging mechanism
+        assert len(out) == 0
+        assert len(err) == 0
         for citation in self.model.citations:
-            assert not (citation.author in out)
-            assert not (citation.doi in out)
-            assert not (citation.author in err)
-            assert not (citation.doi in err)
+            assert not (citation.author in log_output)
```
Comment (petrelharp) [inline tests/test_cli.py:1044]: In figuring out how this works, I discovered we weren't doing what we wanted in this test.

**Comment by petrelharp** [comment]
We should be hitting that test for reals now!

**Comment by petrelharp** [comment]
Okay: pre-commit is failing, you updated the `.pre-commit-config.yaml', and the new version of `black` now requires a newer version of python. Updating the pre-commit versions is not generally something we should do in a PR because then you run the danger of changing a bunch of unrelated stuff (so the PR is messy). However:
- I can't locally run `pre-commit run -a` with the config file on `upstream/main` because it errors for some other obscure reason (maybe this is why you changed the config?)
- the changes it's making with the new versions of black, etc are not bad (AFAICT almost just one extra line after the header in all the files)

So, I'm going to keep the changes to pre-commit; merge this; and deal with pre-commit and the failing macOS CI in a different PR.

**Comment by gregorgorjanc** [comment]
@petrelharp thanks for the help here - yeah, I had to tinker with pre-commit as part of testing ..., which made a mess. Sorry. Thanks for the review and merging this in!

**Pull request merged**